### ISSUE TITLE
Write a migration guide from 2.4.x to 3.0

### ISSUE THREAD

**Issue by maciejwalkowiak** [awspring/spring-cloud-aws#606]

**Comment by youcangetme** [comment]
Yes, please! This is proving rather difficult. This on top of all the Spring 6 issues and my head hurts :(

**Comment by maciejwalkowiak** [comment]
@youcangetme apologies! We should do a better job while releasing milestone versions. Can you tell me what specifically did you find hard/confusing/unclear so that I can make sure to include it?

**Comment by youcangetme** [comment]
Here has been a small sampling of my night. I am sure it is all "pilot error" but its been rather frustrating;

While looking for the missing `QueueMessagingTemplate` I run into this, which would have been nice to know before starting;

> A SqsTemplate should be included in a future milestone simplifying this process.

 Then, with these in my pom;

```
<dependency>
<groupId>io.awspring.cloud</groupId>
<artifactId>spring-cloud-aws-dependencies</artifactId>
<version>3.0.0-M3</version>
<type>pom</type>
</dependency>
<dependency>
<groupId>io.awspring.cloud</groupId>
<artifactId>spring-cloud-aws-starter-sqs</artifactId>
<version>3.0.0-M3</version>
</dependency>
```

I get this error;

```
Failed to read artifact descriptor for io.awspring.cloud:spring-cloud-aws-starter-sqs:jar:3.0.0-M3: Could not find artifact org.springframework.cloud:spring-cloud-build:pom:4.0.0-M4 in central (https://repo.maven.apache.org/maven2)
```
I can't find that library but I can find the 4.0 instead so I use the 3.0.0-M2 instead and everything compiles. When I run the app though;

```
01:21:21.942 [main] ERROR org.springframework.boot.SpringApplication - Application run failed
java.lang.IllegalArgumentException: Unable to instantiate factory class [io.awspring.cloud.autoconfigure.config.parameterstore.ParameterStoreConfigDataLocationResolver] for factory type [org.springframework.boot.context.config.ConfigDataLocationResolver]
	at org.springframework.core.io.support.SpringFactoriesLoader$FailureHandler.lambda$throwing$0(SpringFactoriesLoader.java:650)
...

Caused by: java.lang.IllegalArgumentException: Log types cannot be injected, please use DeferredLogFactory
	at org.springframework.boot.context.config.ConfigDataLocationResolvers.lambda$new$0(ConfigDataLocationResolvers.java:64)

```

So a guide on how to port an existing app would be a real big help. That last one, I am not even using any of that service. I have no idea why that would even be starting up.

I was hoping for a few libs versions getting updated and a few fixes for paths and objects that moved. But right now I have no idea where I went wrong! Argh!

**Comment by vdksoda** [comment]
> Here has been a small sampling of my night. I am sure it is all "pilot error" but its been rather frustrating;
> 
> While looking for the missing `QueueMessagingTemplate` I run into this, which would have been nice to know before starting;
> 
> > A SqsTemplate should be included in a future milestone simplifying this process.
> 
> Then, with these in my pom;
> 
> ```
> <dependency>
> <groupId>io.awspring.cloud</groupId>
> <artifactId>spring-cloud-aws-dependencies</artifactId>
> <version>3.0.0-M3</version>
> <type>pom</type>
> </dependency>
> <dependency>
> <groupId>io.awspring.cloud</groupId>
> <artifactId>spring-cloud-aws-starter-sqs</artifactId>
> <version>3.0.0-M3</version>
> </dependency>
> ```
> 
> I get this error;
> 
> ```
> Failed to read artifact descriptor for io.awspring.cloud:spring-cloud-aws-starter-sqs:jar:3.0.0-M3: Could not find artifact org.springframework.cloud:spring-cloud-build:pom:4.0.0-M4 in central (https://repo.maven.apache.org/maven2)
> ```
> 
> I can't find that library but I can find the 4.0 instead so I use the 3.0.0-M2 instead and everything compiles. When I run the app though;
> 
> ```
> 01:21:21.942 [main] ERROR org.springframework.boot.SpringApplication - Application run failed
> java.lang.IllegalArgumentException: Unable to instantiate factory class [io.awspring.cloud.autoconfigure.config.parameterstore.ParameterStoreConfigDataLocationResolver] for factory type [org.springframework.boot.context.config.ConfigDataLocationResolver]
> 	at org.springframework.core.io.support.SpringFactoriesLoader$FailureHandler.lambda$throwing$0(SpringFactoriesLoader.java:650)
> ...
> 
> Caused by: java.lang.IllegalArgumentException: Log types cannot be injected, please use DeferredLogFactory
> 	at org.springframework.boot.context.config.ConfigDataLocationResolvers.lambda$new$0(ConfigDataLocationResolvers.java:64)
> ```
> 
> So a guide on how to port an existing app would be a real big help. That last one, I am not even using any of that service. I have no idea why that would even be starting up.
> 
> I was hoping for a few libs versions getting updated and a few fixes for paths and objects that moved. But right now I have no idea where I went wrong! Argh!

@youcangetme Came across the same challenge. After reviewing this git repository's pom.xml, I added another maven repository to my local project's pom.xml.

```    
<repositories>
        <repository>
            <id>repository.spring.milestone</id>
            <name>Spring Milestone Repository</name>
            <url>https://repo.spring.io/milestone</url>
        </repository>
</repositories>
```
On rebuilding it was able to find `org.springframework.cloud:spring-cloud-build:pom:4.0.0-M4`. Haven't had a chance to bring up the app and use the functionality yet. I saw your post and decided to share this immediately in case it helps you.

**Comment by youcangetme** [comment]
Thanks @vdksoda. I don't know how I missed that. Get'n old I guess. I will check it out as soon as I can fix the other Spring Framework 6 issues. I have to pick my battles and right now messaging can be left as-is while the rest of this mess gets cleaned up.

**Comment by Jhmmy** [comment]
That would be really helpful indeed. I'm trying to upgrade from 2.4.2 to 3 as part of upgrading an existing application to Spring Boot 3.0.1 and I notice that spring-cloud-aws-messaging is no longer part of the dependencies. As a result we cannot use for example QueueMessagingTemplate. Any suggestions would be really helpful.
Thanks!

**Comment by maciejwalkowiak** [comment]
@Jhmmy there successor for `QueueMessagingTemplate` is in works and will be released in 3.0 RC1 (soon - check PRs to stay up to date). Up until RC1 you must use regular `SqsAsyncClient` unfortunately.

**Comment by montgolfiere** [comment]
JFYI: I have an SQS PoC that I migrated from using Spring Cloud AWS 2.4.4 (on Spring Boot 2.7.6) to Spring Cloud AWS 3.0.0-RC1 (on Spring Boot 3.0.2). In case of `QueueMessagingTemplate` I switched to new `SqsTemplate` (available in 3.0.0-RC1) 👍

**Comment by maciejwalkowiak** [comment]
Closing in favor of https://github.com/awspring/spring-cloud-aws/pull/663.

Everyone is welcome to contribute to a branch from a PR to update migration guide
